// hapi-meal/src/Cooker/Less.js
var _ = require('lodash')
  , fs = require('fs')
  , path = require('path')
  , Hapi = require('hapi')
  , Hasher = require('./../Hasher');

/**
 * Less File Cooker
 * @type {Function}
 */
var Less = module.exports = function Less(app, doneCallback) {
  this._app = app;
  this.settings = this._app.settings.app.assets.less;
  this.doneCallback = doneCallback;
  return this;
};

_.extend(
  Less.prototype,
  {

    cook: function () {
      var self = this;
      try {
        var less = require('less');
        this.parser = new(less.Parser)({
          paths: this.settings.path,
          relativeUrls: false,
          strictImports: false
        });
        this.loopLessFiles(
          function () {
            self.doneCallback();
          }
        );
      } catch (error) {
        console.log(error);
      }
    },

    loopLessFiles: function (callback) {
      this.parser.compiled = 0;
      for (var i = 0, l = this.settings.files.length; i < l; i++) {
        var self = this;
        this.parser.filename = this.settings.files[i];
        var filepath = path.normalize(
          this.settings.path + '/' + this.settings.files[i]
        );
        fs.readFile(
          filepath, 'utf-8',
          function(err, lessContent) {
            try {
              self.parser.compiled++;
              if (err) {
                self.displayError(err);
              } else {
                self.parser.parse(
                  lessContent,
                  function (err, tree) {
                    if (err) {
                      self.displayError(err);
                    } else {
                      try {
                        self.registerCompiledFile(
                          tree.toCSS({
                            compress: true,
                            yuicompress: true
                          }),
                          filepath,
                          self.parser.filename
                        );
                      } catch (err) { self.displayError(err); }
                    }
                    if (self.parser.compiled === l
                      && typeof callback === 'function') {
                      callback();
                    }
                  }
                );
              }
            } catch (err) { self.displayError(err); }
          }
        );
      }
    },

    /**
     * Register compiled files as app properties
     * TODO : Check redis or other kind of storage
     * @param fileContent
     * @param filePath
     * @param fileName
     */
    registerCompiledFile: function(fileContent, filePath, fileName) {
      var hash = new Hasher().sumString(
        filePath,
        fileContent
      );
      console.log('Registering LESS file "' + fileName + '" (hash "' + hash + '")');
      this._app.resources.css.content[hash] = fileContent;
      this._app.resources.css.path[fileName] = hash;
    },

    displayError: function(error) {
      console.log(
        "Error packing LESS",
        {
          message: error.message,
          line: error.line,
          file: error.filename
        }
      );
    }
  }
);
